### PR TITLE
[MRG] MinHash class refactoring

### LINKED ISSUE #82: Possible inconsistency of methods of class `MinHash`
The class defines a `__len__()`, but the length of what is returned by `get_mins` is not necessarily consistent:

```python
In [1]: import sourmash_lib._minhash

In [2]: mh = sourmash_lib._minhash.MinHash(1000, 21)

In [3]: len(mh)
Out[3]: 1000

In [4]: mh.get_mins()
Out[4]: []
```

### LINKED ISSUE #284: Use iterators instead of directly accessing `mins` and `abunds` from C++ classes
Punted from #282:

- stop using `mins` directly. Since virtual inheritance only works for methods (and not members), this was creating a lot of weird boilerplate in Cython. 
  - The stopgap measure is to call `abunds` when using `KmerMinAbundance`, but this is still kinda ugly.
  - In the future, use iterators to access `hashes`, `abundances`, or pairs of `(hash, abundance)`. `count_common` doesn't need to be implemented in `KmerMinAbundance` if we do this, since the base `KmeMinHash::count_common` can iterate over the hashes only.
  - Expose iterators for Cython, and avoid accessing `mins` and `abunds` directly.

### LINKED ISSUE #611: add useful methods from sig submodule into MinHash() objects
The functionality to flatten and downsample signatures that is in `sourmash/sig/__main__.py` as of #587 could usefully be added to `MinHash(...)` objects.

### LINKED ISSUE #618: Should downsample_* methods be inplace (instead of returning a new copy)?
https://github.com/dib-lab/sourmash/blob/314bd8e06025d0fd22c37fced0fa5f22aa832a4f/sourmash/_minhash.pyx#L248-L291

It's better to make a explicit copy if needed than do a new copy by default.

@ctb comment: " got a little bit wary of modifying minhashes in place a while back after running into some problems with md5sums changing"

### LINKED ISSUE #896: should MinHash objects support list or dict interface?
We could support iteration and/or indexing. e.g.

```
for hash in mh_object:
   ...
```

and

```
for hash in mh_object:
    abund = mh_object[hash]
```

### LINKED ISSUE #951: deprecate is_molecule_type for sourmash 4
per #936

### PR DESCRIPTION
**Pull request by ctb**
Many minor refactors and associated deprecations of the `MinHash` class.

Addresses some of #338 
Fixes #611 by adding `flatten`
Fixes #82, fixes #284, and fixes #896 by providing a `.hashes` attribute that implements a read-only dictionary interface to hashes
Fixes #951 by deprecating `is_molecule_type`.
Fixes #618 by settling on `downsample(...)` as the downsampling API. 

Relevant to general MinHash API refactoring #720, #885 #999 but doesn't address core issues there.

TODO before merge:
- [ ] maybe document that only functions exposed at sourmash top level are public API under semantic versioning...
- [ ] put together some doctests for the top level API / write docs and doctests against `MinHash` since it's darn useful
- [x]  deprecate `translate_codon` or remove from MinHash class at any rate; @luizirber said "We can have a submodule for working with these translations, I think it is useful for prototyping new ideas."
- [ ] summarize proposed `__init__` refactoring in a new issue
---

- [x] Is it mergeable?
- [x] `make test` Did it pass the tests?
- [x] `make coverage` Is the new code covered?
- [x] Did it change the command-line interface? Only additions are allowed
  without a major version increment. Changing file formats also requires a
  major version number increment.
- [x] Was a spellchecker run on the source code and documentation after
  changes were made?

# original comment - Pedestrian MinHash class refactoring for 4.0?

This is a trial cleanup & regularization of the MinHash class, to see how it all ...looks and feels. Comments welcome!

related issues:

https://github.com/dib-lab/sourmash/issues/720
https://github.com/dib-lab/sourmash/issues/82
https://github.com/dib-lab/sourmash/issues/338
https://github.com/dib-lab/sourmash/issues/611
https://github.com/dib-lab/sourmash/issues/896
https://github.com/dib-lab/sourmash/issues/999

## Constructor refactoring

see specific @luizirber proposal in [#999](https://github.com/dib-lab/sourmash/issues/999#issuecomment-633707771).

I'm kind of in favor of
* not doing a hash function switch over for 4.0
* requiring ksize and moltype
* requiring one of num or scaled;
* (and changing `n` -> `num`)

## changes to methods

(deprecate in 3.x, remove in 4.0.)

- [x] deprecate one or both of `get_hashes` (?) or `get_mins` , maybe in favor of a view object/property (see below) like `.hashes`?
- [x] remove `subtract_mins`
- [x] add` -> `add_kmer`?
- [x] remove update
- [x] deprecate `translate_codon` or remove from MinHash class at any rate
- [x] rename `downsample_n` to `downsample_num`, or `downsample(num=...)`
- [x] eliminate `max_hash` stuff completely from public API, in favor of `scaled`
- [x] eliminate `mins` completely from public API, in favor of `hash` and `hashes`
- [x] eliminate `is_molecule_type`

## Add:

* `flatten` to remove abundances.
- [x] what about adding a view object for the hashes that provides a list/dict interface? e.g. [#896](https://github.com/dib-lab/sourmash/issues/896)

### CODE REVIEW

**Comment by ctb** [comment]
I think this might become close to mergable in 3.x if we deprecate rather than remove various methods, such as `get_mins` and `subtract_mins`, and also add a few tests to make sure that `get_mins()` and so on still work. Thoughts, @luizirber ?

**Comment by luizirber** [comment]
> I think this might become close to mergable in 3.x if we deprecate rather than remove various methods, such as `get_mins` and `subtract_mins`, and also add a few tests to make sure that `get_mins()` and so on still work. Thoughts, @luizirber ?

That would be awesome! Deprecating in 3.5 and removing in 4 is kind of short notice, but way better than just erroring. Still think we need a transition doc, but it can be way shorter ("check for DeprecationWarning!")

**Comment by luizirber** [comment]
> inspired by @luizirber talk, I decided to try my own experimental PR!

Yay experimental PRs! Even if it is not all merged, we can pick most bits and pieces and add in other PRs =] 

> This can't be merged as-is because it breaks the API in several places, but I can see keeping the tests as they are and adding deprecations... thoughts welcome!
> # Pedestrian MinHash class refactoring for 4.0?

> I'm kind of in favor of
> 
>     * not doing a hash function switch over for 4.0
>     * requiring ksize and moltype
>     * requiring one of num or scaled;
>     * (and changing `n` -> `num`)

Yeah, I think the hash function switch on `__init__` should come later (but we should start using them internally)

> deprecate one or both of `get_hashes` (?) or `get_mins` , maybe in favor of a view object/property (see below) like `.hashes`?

I really like the view style!

> add`->`add_kmer`?

> remove update

I think the goal was to have something more `dict`/`set` like? But I don't remember ever using it.

> deprecate `translate_codon` or remove from MinHash class at any rate

We can have a submodule for working with these translations, I think it is useful for prototyping new ideas.

> rename `downsample_n` to `downsample_num`, or `downsample(num=...)`

I kinda like `downsample(num=...)`
 
> eliminate `max_hash` stuff completely from public API, in favor of `scaled`
:+1: 

> eliminate `mins` completely from public API, in favor of `hash` and `hashes`
:100: 

> what about adding a view object for the hashes that provides a list/dict interface? e.g. [#896](https://github.com/dib-lab/sourmash/issues/896)
I really like this style.

**Comment by ctb** [comment]
> > I think this might become close to mergable in 3.x if we deprecate rather than remove various methods, such as `get_mins` and `subtract_mins`, and also add a few tests to make sure that `get_mins()` and so on still work. Thoughts, @luizirber ?
> 
> That would be awesome! Deprecating in 3.5 and removing in 4 is kind of short notice, but way better than just erroring. Still think we need a transition doc, but it can be way shorter ("check for DeprecationWarning!")

I think we can also let many things persist (`get_mins()`, etc.) past 4.0 without much harm. It's mostly the complete incompatibilities that I want to remove!

File: tests/test__minhash.py [hunk 1]
```diff
@@ -230,39 +234,26 @@ def test_size_limit(track_abundance):
...
+    scaled = _get_scaled_for_max_hash(35)
+    print('XX', scaled, _get_max_hash_for_scaled(scaled))
     mh = MinHash(0, 4, track_abundance=track_abundance, scaled=scaled)
     assert mh.max_hash == 35
 
     mh.add_hash(10)
     mh.add_hash(20)
     mh.add_hash(30)
-    assert mh.get_mins() == [10, 20, 30]
+    assert mh.hashes == [10, 20, 30]
```
Comment (ctb) [inline tests/test__minhash.py]: hmm a bit surprised this still works, shouldn't it be list(mh.hashes)?

**Comment by ctb** [comment]
Misc thoughts for putting this up for merge --

* `_minhash` was always an internal API name, `sourmash.MinHash` is what should have been used. So I am keeping the module rename 👍 
- [ ] we'll deprecate (but keep) `get_mins`, while switching to using `hashes`.
- [ ] `get_max_hash_for_scaled` and `get_scaled_for_max_hash` were not exposed at top level and were part of the `_minhash` module, changing their names should be fine.
- [ ] speaking of which we should maybe document that only functions exposed at `sourmash` top level are public API under semantic versioning...
- [ ] deprecate `MinHash` methods `is_molecule_type`, `downsample_n` and `downsample_scaled`
- [ ] implement `downsample(num/scaled=...)`
- [ ] deprecate `add` and `update` in favor of `add_kmer` and `set_abundances(..., clear=False)`
- [ ] deprecate `downsample_max_hash`
- [ ] deprecate `max_hash` property (?) or make it internal only (`_max_hash`)
- [ ] add tests `test_max_hash_and_scaled_error` and `test_mh_subtract` back in, for now, with a note to maybe remove them later
- [ ] misc thought: maybe `hashes` could (for some future `MinHash` class be read-write? for now we should make it a read-only dict interface.

**Comment by ctb** [comment]
maaaybe think about which other parts of the top level API to deprecate. and/or put together some doctests for the top level API.

**Comment by ctb** [comment]
hmm also make issue to refactor Rust API accordingly - e.g. `max_hash` conversion.

**Comment by ctb** [comment]
- [ ] look at dna/protein/hp/dayhoff properties; are they used anywhere?
- [ ] add tests for deprecated features like `get_mins()`

File: tests/test__minhash.py (hunk 1 above)
Comment (ctb) [reply tests/test__minhash.py]: indeed, it should be. this test was masked by another one below... and didn't work :)

fixed in 9d178c950fb45

**Comment by ctb** [comment]
more observations/TODO:

- [ ] `add_kmer` should check that sequence length is exactly ksize
- [ ] add `flatten`.
- [ ] re-add some (minimal) tests for deprecated features like `get_mins()` etc.
- [ ] make issue to refactor Rust API accordingly - e.g. max_hash conversion.
- [ ] examine dna/protein/hp/dayhoff properties for deprecation

also still need to take care of `translate_codon`. I'd also like to:

- [ ] write docs and doctests against `MinHash` since it's darn useful, but I might punt that.

**Comment by ctb** [comment]
ok, we have some fun inconsistencies in the molecule type -- probably not to be resolved here, but we use `MinHash.is_dna` and `MinHash.is_protein` vs `MinHash.hp` and `MinHash.dayhoff`.

Maybe we should make moltype be an explicit argument to the MinHash constructor, i.e. `moltype="DNA"`. and/or use an enum...

**Comment by ctb** [comment]
any thoughts on the code changes, @luizirber ?

**Comment by luizirber** [comment]
> any thoughts on the code changes, @luizirber ?

Really like it, but... even tho `sourmash._minhash` was internal, it is quite possible it is going to break working scripts based on `3.x`. Maybe keep an alias from `sourmash.minhash` to `sourmash._minhash` before 4.0, and then remove the alias in 4.0?

**Comment by ctb** [comment]
> > any thoughts on the code changes, @luizirber ?
> 
> Really like it, but... even tho `sourmash._minhash` was internal, it is quite possible it is going to break working scripts based on `3.x`. Maybe keep an alias from `sourmash.minhash` to `sourmash._minhash` before 4.0, and then remove the alias in 4.0?

hmm, ok? underscores are typically indicative of private modules, but you are certainly right that I and others probably used `_minhash` directly. tsk tsk. :)

**Comment by ctb** [comment]
(ironically, in my dev environment, I've been running into exactly this with spacegraphcats and sourmash interactions. sigh.)

**Comment by ctb** [comment]
ok, added in 8d3c083

I went down a bit of a rabbit hole [reading about warnings](https://docs.python.org/3/library/warnings.html), and I ended up using `FutureWarning` instead of `DeprecationWarning` so that it would show up when people were importing the new module. (It works fine, I tested it on spacegraphcats.)

We may want to consider our instructions for migrating to 4.x to include running scripts with the `-W error` flag to trigger the `DeprecationError` => errors.

In any case - this is ready for review again :)

**Comment by luizirber** [comment]
> ok, added in [8d3c083](https://github.com/dib-lab/sourmash/commit/8d3c083af123e824baf5960b397a575193bdabd9)
> 
> I went down a bit of a rabbit hole [reading about warnings](https://docs.python.org/3/library/warnings.html), and I ended up using `FutureWarning` instead of `DeprecationWarning` so that it would show up when people were importing the new module. (It works fine, I tested it on spacegraphcats.)

Nice! We could also use [pytest warning capture](https://docs.pytest.org/en/latest/warnings.html) to see if they are actually being thrown, but maybe that's too much.

> We may want to consider our instructions for migrating to 4.x to include running scripts with the `-W error` flag to trigger the `DeprecationError` => errors.

And also suggest people run CI/tests with [`PYTHONDEVMODE=1`](https://github.com/dib-lab/sourmash/blob/16e6934f9ff707ec81a453856ccfc69bd17009dd/.travis.yml#L33), so they will show up more explicitly?

> In any case - this is ready for review again :)

I think there isn't anything else that is breaking compatibility, but would like to take a closer look before approving. Soon. =]

**Comment by ctb** [comment]
heh, the last commit was necessary because of a fun chain of events:

* once we tagged this as 4.0, the DeprecationWarning for `max_hash` was triggered.
* this DeprecationWarning then interfered with a test that was explicitly NOT expecting a warning.

**Comment by luizirber** [comment]
Is this PR also covering #1145 now?

In any case, we might need a PR derived from this one for the `DeprecationWarning`s and merge it in `stable`? (and then release it as `3.4.2` and continue updating `latest`? =])

**Comment by ctb** [comment]
actually, this one should probably be merged into stable (and released), then
merged into latest, and then I should construct new PRs against latest to
remove the features to be removed in 4.0.

whee. :)

File: sourmash/minhash.py [hunk 2]
```diff
@@ -0,0 +1,675 @@
...
+                current_version=VERSION,
+                details='Use scaled instead.')
+    def downsample_max_hash(self, *others):
+        """Copy this object and downsample new object to min of ``*others``.
+
+        Here, ``*others`` is one or more MinHash objects.
+        """
+        max_hashes = [x.max_hash for x in others]
+        new_max_hash = min(self.max_hash, *max_hashes)
+        new_scaled = get_scaled_for_max_hash(new_max_hash)
```
Comment (luizirber) [inline sourmash/minhash.py]:
```suggestion
        new_scaled = _get_scaled_for_max_hash(new_max_hash)
```
(this is a bug that [sonarcloud](https://sonarcloud.io/project/issues?id=dib-lab_sourmash&open=AXOQyD9y8hpWe-mDGL3C&pullRequest=1128&resolved=false&types=BUG) caught)

**Comment by ctb** [comment]
> Is this PR also covering #1145 now?

Finally understood this comment, just a bit too late. Sigh. Yes, I merged in #1145 because I got tired of resolving conflicts...  I didn't update the comment to reflect that.

**Pull request merged**